src/cgame/rocket/lua/Player.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+===========================================================================
+
+Daemon GPL Source Code
+Copyright (C) 2012 Unvanquished Developers
+
+This file is part of the Daemon GPL Source Code (Daemon Source Code).
+
+Daemon Source Code is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Daemon Source Code is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Daemon Source Code.  If not, see <http://www.gnu.org/licenses/>.
+
+In addition, the Daemon Source Code is also subject to certain additional terms.
+You should have received a copy of these additional terms immediately following the
+terms and conditions of the GNU General Public License which accompanied the Daemon
+Source Code.  If not, please request a copy in writing from id Software at the address
+below.
+
+If you have questions concerning this license or the applicable additional terms, you
+may contact in writing id Software LLC, c/o ZeniMax Media Inc., Suite 120, Rockville,
+Maryland 20850 USA.
+
+===========================================================================
+*/
+
+#include "Player.h"
+#include "../../cg_local.h"
+
+using Shared::Lua::RegType;
+using Shared::Lua::LuaLib;
+
+namespace {
+
+class Player {};
+
+static Player player;
+
+#define GETTER(name) { #name, Get##name }
+
+#define GET_FUNC( name, var, type ) \
+static int Get##name( lua_State* L ) \
+{ \
+	playerState_t* ps = &cg.snap->ps; \
+	if ( ps ) \
+	{ \
+		lua_push##type( L, var ); \
+		return 1; \
+	} \
+	return 0; \
+}
+
+GET_FUNC( team, BG_TeamName( ps->persistant[ PERS_TEAM ] ), string )
+GET_FUNC( class, BG_Class( ps->stats[ STAT_CLASS ] )->name, string )
+GET_FUNC( weapon, BG_Weapon( ps->stats[ STAT_WEAPON ] )->name, string )
+GET_FUNC( hp, ps->stats[ STAT_HEALTH ], integer )
+GET_FUNC( ammo, ps->ammo, integer )
+GET_FUNC( clips, ps->clips, integer )
+GET_FUNC( credits, ps->persistant[ PERS_CREDIT], integer )
+GET_FUNC( score, ps->persistant[ PERS_SCORE], integer )
+
+}  // namespace
+
+namespace Shared {
+namespace Lua {
+
+template<> void ExtraInit<Player>(lua_State* /*L*/, int /*metatable_index*/) {}
+
+RegType<Player> PlayerMethods[] =
+{
+	{ nullptr, nullptr },
+};
+
+luaL_Reg PlayerGetters[] =
```

Comment:
I tested and `score` always gives the result for the real client, not the followed player. Maybe that one should be removed

Reply:
I think it's still useful even if it doesn't follow the player for whatever reason...

Reply:
How about making it like a static method instead of an instance method of player? Meaning it would be called like `Player.score` instead of `player.score`. The Rmlui function `template<> void ExtraInit<DataFormatter>(lua_State* L, int metatable_index)` has an example

Reply:
The `player` struct is essentially static. It's just syntactic sugar around what you've done in the other Lua structs.

The reason that score doesn't change is because the game only stores the score and ping in PERS_SCORE and PERS_PING so if we lose it when we specate a player, then we'd maintain their score.

I think it's honestly more confusing to have `player` and `Player` be different. I think we should either accept this as a quirk of the game or fix the sgame to also set PERS_SCORE to the spectated player's score and score the real player's score elsewhere.

Reply:
> The `player` struct is essentially static. It's just syntactic sugar around what you've done in the other Lua structs.

I get it, but we can pretend it's an object referring to the currently playing or spectated player. In the future maybe the Player class could be extended to work for other players too, at least for the fields knowable from their entityState_t. So I think it's bad to mix APIs that act on the real client with ones that act on the spectated player. `player` vs. `Player` is indeed a pretty bad way to differentiate these two kinds of functions but I suggested it as being easier than creating a new object type or whatever.



> fix the sgame to also set PERS_SCORE to the spectated player's score and score the real player's score elsewhere.

I believe this would break the server as is uses PERS_SCORE when generating the player list for serverinfo queries.

Reply:
Ok, I check if we're spectating someone and report their score instead with the caveat that their score might be slightly delayed based on CG_RequestScores.